### PR TITLE
Created use case for find Campaigns by MetaInfo/Id/StartDate

### PR DESCRIPTION
**Pull request by 04tudor**

### CODE REVIEW

File: application/pom.xml [hunk 1]
```diff
@@ -33,5 +33,13 @@
             <artifactId>equalsverifier</artifactId>
             <scope>test</scope>
         </dependency>
+
+        <dependency>
+            <groupId>org.assertj</groupId>
+            <artifactId>assertj-core</artifactId>
+            <version>3.25.1</version>
```
Comment (DemiusAcademius) [inline application/pom.xml:40]: we have a practice indicate all version numbers in main pom.xml and use dependency management

File: application/src/main/java/md/maib/retail/CampaignAllInfo.java [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+import md.maib.retail.model.campaign.CampaignId;
+import md.maib.retail.model.campaign.CampaignMetaInfo;
+import md.maib.retail.model.campaign.CampaignState;
+import md.maib.retail.model.campaign.LoyaltyEventType;
+import md.maib.retail.model.conditions.Rule;
+import org.threeten.extra.Interval;
+
+import java.util.Collection;
+@Getter
+public class CampaignAllInfo {
```
Comment (DemiusAcademius) [inline application/src/main/java/md/maib/retail/CampaignAllInfo.java]: you can use @RequiredArgsConstructor

File: application/src/main/java/md/maib/retail/CampaignMapper.java [hunk 3]
```diff
@@ -0,0 +1,29 @@
+package md.maib.retail;
+
+import md.maib.retail.model.campaign.*;
+import md.maib.retail.model.conditions.Rule;
+import org.threeten.extra.Interval;
+
+import java.util.Collection;
+
+public class CampaignMapper {
+    public static CampaignSomeInfo mapToCampaignSomeInfo(Campaign campaign) {
```
Comment (DemiusAcademius) [inline application/src/main/java/md/maib/retail/CampaignMapper.java]: You can use static factory in Campaign xxx Info classes

File: application/src/main/java/md/maib/retail/CampaignAllInfo.java [hunk 4]
```diff
@@ -0,0 +1,34 @@
+package md.maib.retail;
```
Comment (DemiusAcademius) [inline application/src/main/java/md/maib/retail/CampaignAllInfo.java]: main package should be md.maib.retail.application

Comment (04tudor) [reply application/src/main/java/md/maib/retail/CampaignAllInfo.java]: but it is in module application.
i should rename and package?

File: application/src/test/java/md/maib/retail/services/FindCampaignByIdServiceTest.java [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+import md.maib.retail.model.campaign.CampaignId;
+import md.maib.retail.model.campaign.CampaignState;
+import md.maib.retail.model.ports.Campaigns;
+import org.junit.jupiter.api.Test;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+@UnitTest
+public class FindCampaignByIdServiceTest {
+    Campaigns campaigns;
```
Comment (DemiusAcademius) [inline application/src/test/java/md/maib/retail/services/FindCampaignByIdServiceTest.java]: How you mock this port?

may be use org.mockito.Mockito?

annotation @Mock on Campaigns campaigns

special methods
@BeforeEach
    void setupProcessor() {
        target = findCampaignsByIdService(campaigns);
    }

    @AfterEach
    void tearDown() {
        verifyNoMoreInteractions(campaigns);
    }

and in test methods something like that
when(campaigns.findActive(any())).thenReturn(Optional.of(campaign));

Comment (04tudor) [reply application/src/test/java/md/maib/retail/services/FindCampaignByIdServiceTest.java]: done

File: application/src/main/java/md/maib/retail/application/CampaignMapper.java [hunk 6]
```diff
@@ -0,0 +1,29 @@
+package md.maib.retail.application;
+
+import md.maib.retail.model.campaign.*;
+import md.maib.retail.model.conditions.Rule;
+import org.threeten.extra.Interval;
+
+import java.util.Collection;
+
+public class CampaignMapper {
+    public static CampaignSomeInfo mapToCampaignSomeInfo(Campaign campaign) {
```
Comment (DemiusAcademius) [inline application/src/main/java/md/maib/retail/application/CampaignMapper.java]: Each result / dto class can contain static factory for construct it from Campaign (named as of or valueOf)
Also we can use simple constructor

File: application/src/main/java/md/maib/retail/application/find_campaign_by_metainfo/FindCampaignByMetaInfoService.java [hunk 7]
```diff
@@ -0,0 +1,28 @@
...
+
+ public class FindCampaignByMetaInfoService implements FindCampaignByMetaInfoUseCase{
+    private final Campaigns campaigns;
+
+    public FindCampaignByMetaInfoService(Campaigns campaigns) {
+        this.campaigns = Objects.requireNonNull(campaigns,"Campaigns must not be null");
+    }
+
+    @Override
+    public Optional<CampaignAllInfo> findByMetaInfo(String key, String value) {
```
Comment (DemiusAcademius) [inline application/src/main/java/md/maib/retail/application/find_campaign_by_metainfo/FindCampaignByMetaInfoService.java]: findByMetaInfo can return many campaigns.
Example: all campaigns of Felicia: by brands Farmacia and Optica, standard and Promo

File: application/src/test/java/md/maib/retail/application/services/CampaignsListByDateServiceTest.java [hunk 8]
```diff
@@ -0,0 +1,72 @@
...
+
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.mockito.Mockito.verifyNoMoreInteractions;
+import static org.mockito.Mockito.when;
+
+@ExtendWith(MockitoExtension.class) public class CampaignsListByDateServiceTest {
+    @Mock
+    Campaigns campaigns;
+
+    @InjectMocks
```
Comment (DemiusAcademius) [inline application/src/test/java/md/maib/retail/application/services/CampaignsListByDateServiceTest.java]: Why InjectMocks? we create instance in setup() method

File: application/src/test/java/md/maib/retail/application/services/CampaignsListByDateServiceTest.java [hunk 9]
```diff
@@ -0,0 +1,72 @@
...
+    CampaignsListByDateService campaignsListByDateService;
```
Comment (DemiusAcademius) [inline application/src/test/java/md/maib/retail/application/services/CampaignsListByDateServiceTest.java]: Use Interface, not implementation

File: domain-model/src/test/java/md/maib/retail/model/RuleTest.java [hunk 10]
```diff
@@ -52,7 +51,7 @@ void testRule() {
         // Rule Assertions
         assertThat(rule.getId()).isNotNull();
         assertThat(rule.getConditions()).containsExactly(condition);
-        assertThat(rule.getEffects()).containsExactly(effect);
+        Assertions.assertThat(rule.getEffects()).containsExactly(effect);
```
Comment (DemiusAcademius) [inline domain-model/src/test/java/md/maib/retail/model/RuleTest.java:54]: Rule is a entity. It has state and can be changed. For equality of Rule we use ID, and EqualsVerifier will be preferred

File: application/src/main/java/md/maib/retail/application/find_campaign_by_metainfo/FindCampaignByMetaInfoService.java (hunk 7 above)
Comment (04tudor) [reply application/src/main/java/md/maib/retail/application/find_campaign_by_metainfo/FindCampaignByMetaInfoService.java]: i should make as a list of campaigns?

**Pull request merged**